lib/usd/translators/meshWriter_BlendShapes.cpp

```diff
@@ -0,0 +1,977 @@
+        // out what the _actual_ base mesh was.
+        MayaBlendShapeDatum info = {};
+        info.blendShapeDeformer = curBlendShape;
+        info.outputGeomIndex = outputGeomPlugIdx;
+        fnGeoFilter.setObject(curBlendShape);
+        MObject inputGeo = fnGeoFilter.inputShapeAtIndex(outputGeomPlugIdx, &stat);
+        CHECK_MSTATUS_AND_RETURN_IT(stat);
+        MObject outputGeo = fnGeoFilter.outputShapeAtIndex(outputGeomPlugIdx, &stat);
+        CHECK_MSTATUS_AND_RETURN_IT(stat);
+        info.baseMesh = inputGeo;
+        info.deformedMesh = deformedMesh;
+
+        fnBlendShape.setObject(curBlendShape);
+        info.numWeights = fnBlendShape.numWeights();
+        MIntArray weightIndices;
+        stat = fnBlendShape.weightIndexList(weightIndices);
+        CHECK_MSTATUS_AND_RETURN_IT(stat);
+
+        // NOTE: (yliangsiew) Ok, so for each weight, need to go targetItemIndexList()
+        // for each base object, then use the base object outputGeometry logicalIndex and
+        // assume that it is the same logical index for the inputTarget logical index; this
+        // is the way (that we will find the inputTargetItem plug to read component data from)
+        MPlug plgInTgts = fnBlendShape.findPlug(kMayaAttrNameBlendShapeInTgt, false, &stat);
+        CHECK_MSTATUS_AND_RETURN_IT(stat);
+        MPlug plgInTgt = plgInTgts.elementByLogicalIndex(outputGeomPlugIdx, &stat);
+        CHECK_MSTATUS_AND_RETURN_IT(stat);
+
+        // NOTE: (yliangsiew) So after we call targetItemIndexList which associates a given weight
+        // index and a base object, we can infer that the inputTarget group logical index matches
+        // that of the outputGeometry logical index (i.e. the base object).
+        // The logical index of the weight plug should match that of the
+        // inputTargetGroup that is being driven by said weight. (Confirmed by @williamkrick from
+        // ADSK).
+        MPlug plgInTgtGrps
+            = UsdMayaUtil::FindChildPlugWithName(plgInTgt, kMayaAttrNameBlendShapeInTgtGrp);
+        TF_VERIFY(!plgInTgtGrps.isNull());
+
+        // NOTE: (yliangsiew) Problem: looks like there's a maya bug where you have to twiddle the
+        // blendshape weight directly before these kComponentListData-type plugs get evaluated.
+#if MAYA_BLENDSHAPE_EVAL_HOTFIX
+        mayaBlendShapeTriggerAllTargets(curBlendShape);
+#endif
+
+        for (unsigned int i = 0; i < weightIndices.length(); ++i) {
+            MayaBlendShapeWeightDatum weightInfo = {};
+            weightInfo.weightIndex = weightIndices[i];
+            weightInfo.inputTargetIndex = outputGeomPlugIdx;
+
+            stat = fnBlendShape.targetItemIndexList(
+                weightInfo.weightIndex, outputGeo, weightInfo.targetItemIndices);
+            CHECK_MSTATUS_AND_RETURN_IT(stat);
+
+            unsigned int numInTgtGrps = plgInTgtGrps.numElements();
+            if (numInTgtGrps == 0) {
+                continue;
+            }
```

Comment:
I think what you have written here will work, but please test a scene were you have a weight value for a non-existent target group. Maya can handle that case (by ignoring the entry which I think you are successfully doing here) so there could be scenes out there with more weights then targets.